subprojects/plugins/src/main/resources/org/gradle/api/internal/plugins/unixStartScript.txt

```diff
@@ -17,37 +17,145 @@
+#       might as well cd to the directory containing gradlew and run it as
+#
+#           ./gradlew
+#
+#       Any arguments you include on the command line will be passed through to
+#       Java after the classname, to be available as parameters to main().
+#
+#   (2) This script is generated from the Groovy template file
+#       subprojects/plugins/src/main/resources/org/gradle/api/internal/plugins/unixStartScript.txt
+#       within the Gradle project, so that's where you need to submit any
+#       patches.
+#
+#       Gradle is a meta-build system used by the project that you're building
+#       or installing. It's like autoconf but for projects that are written in
+#       Java and related languages. It's also used to build parts of the Gradle
+#       project itself.
+#
+#       Gradle is mostly written in Java and Groovy, but there are wrapper
+#       scripts for various environments, including this one for POSIX shells.
+#<% /*
+#       ... and if you're reading this, this IS the template just mentioned.
+#
+#       The Groovy template language is run in two phases.
+#
+#        1. Any character following \ is passed unmodified through to the
+#           next phase, while the \ is removed. Any other $ followed by
+#           varName or {varName} is replaced by the value of that variable.
+#
+#        2. The result of the first phase is parsed and run in a similar
+#           manner to JSP or MASON or PHP: anything within < % ... % > is a
+#           code block, anything else is sent as output, subject to the
+#           flow imposed by any code segments.
+#
+#           Yes, if you need a \ in the second phase, you need to write \\
+#
+#       See https://docs.groovy-lang.org/next/html/documentation/ section §3.15
+#       (Template Engines) for details.
+#
+#       Bear in mind that the "output" is a POSIX shell script, which has its
+#       own ideas about escaping with backslashes, so to get «\» you need to
+#       write «\\\\» and to get «$» you need to write «\\\$».
+#
+#       Code segments are used in this script to control conditional inclusions:
+#           (a) to build a module rather than an app, indicated by passing
+#               '--module XYZ' in place of the classname that contains main
+#           (b) to optionally define a ${appNameSystemProperty} symbol
+#
+#       These are the variables used in this script, and the defaults when
+#       building Gradle itself:
+#
+#           ${appNameSystemProperty}    org.gradle.appname
+#           ${applicationName}          Gradle
+#           ${mainClassName}            org.gradle.wrapper.GradleWrapperMain
+#           ${optsEnvironmentVar}       GRADLE_OPTS
+#           ${defaultJvmOpts}           -Dfile.encoding=UTF-8 "-Xmx64m" "-Xms64m"
+#
+#       (These are all specified in
+#       subprojects/build-init/src/main/java/org/gradle/api/tasks/wrapper/Wrapper.java
+#       within the Gradle project; they may be different for other projects.)
+# */ %>
```

Comment:
I would take some variation of what you have here between <%/%> and move it to subprojects/plugins/src/main/java/org/gradle/api/internal/plugins/UnixStartScriptGenerator.java

That's the class that uses this template to generate the script. If you think it would be helpful for future readers, we could add a link back to UnixStartScriptGenerator from the template that is not included in the generated script. My main concern with packing too much developer documentation in the template itself is that it's already confusing and unnatural to escape everything.

Reply:
I take your point on size, but this is only visible to someone reading the template, at which point it's critically important that they understand the three-phase parsing so I'd rather leave that part in place.

On the other hand, they probably don't need the list of available Groovy variables unless they're making substantial changes, in which case a link elsewhere would suffice.

I assume some kind of JavaDoc format would be appropriate? I'm very rusty on what that would entail. Perhaps that should be a separate PR?